include/sparrow/any_data.hpp

```diff
@@ -0,0 +1,349 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <any>
+#include <ranges>
+#include <tuple>
+#include <type_traits>
+#include <typeindex>
+#include <vector>
+
+#include "sparrow/any_data_utils.hpp"
+
+namespace sparrow
+{
+    /// A class that can own or not any object and expose it as a raw pointer.
+    template <typename T>
+    class any_data
+    {
+    public:
+
+        explicit any_data(T* data);
+
+        template <mpl::unique_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+        template <mpl::shared_ptr_or_derived U>
+            requires std::is_same_v<typename U::element_type, T>
+        explicit any_data(U data);
+
+
+        explicit any_data(T data)
+            requires(!mpl::smart_ptr_and_derived<T>);
+
+        T* get();
+        const T* get() const;
+
+        // Returns the data as a reference to the type U.
+        // The type U must be the same as the type of the data. Otherwise std::bad_any_cast exception is
+        // thrown.
+        template <class U>
+        U get_data();
+
+        template <class U>
+        const U get_data() const;
+
+        [[nodiscard]] bool owns_data() const noexcept;
+
+        std::type_index type_id() const;
+
+    private:
+
+        std::any m_owner;
+        T* m_raw_ptr = nullptr;
+    };
+
+    template <typename T>
+    any_data<T>::any_data(T* data)
+        : m_raw_ptr(data)
+    {
+    }
+
+    template <typename T>
+    template <mpl::unique_ptr_or_derived U>
+        requires std::is_same_v<typename U::element_type, T>
+    any_data<T>::any_data(U data)
+        : m_owner(std::move(value_ptr<typename U::element_type, typename U::deleter_type>(std::move(data))))
```

Comment:
The first move can be removed, the anonymous value_ptr is an rvalue.

Reply:
Done